pkg/sentry/devices/memdev/kmsg.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memdev
+
+import (
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/buffer"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const (
+	// BufferEntryMax is the maximum number of entries retained by kmsg before new entries overwrite old ones.
+	BufferEntryMax = 512
+
+	// BufferMax is the maximum size of an individual entry, in bytes.
+	BufferMax = 1024
+)
+
+const kmsgDevMinor = 11
+
+// kmsgDevice implements vfs.Device for /dev/kmsg.
+type kmsgDevice struct{}
+
+// Open implements vfs.Device.Open.
+func (kmsgDevice) Open(ctx context.Context, mnt *vfs.Mount, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	creds := auth.CredentialsFromContext(ctx)
+	k := kernel.KernelFromContext(ctx)
+	if !creds.HasCapabilityIn(linux.CAP_SYSLOG, k.RootUserNamespace()) && !creds.HasCapabilityIn(linux.CAP_SYS_ADMIN, k.RootUserNamespace()) {
+		return nil, syserror.EPERM
+	}
+	fd := &kmsgFD{}
+	if err := fd.vfsfd.Init(fd, opts.Flags, mnt, vfsd, &vfs.FileDescriptionOptions{
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// kmsgFD implements vfs.FileDescriptionImpl for /dev/kmsg.
+type kmsgFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.NoLockFD
+
+	// buffer is a circular buffer that store user's write to kmsg.
+	m          sync.Mutex
+	buffer     [BufferEntryMax]*buffer.View
```

Comment:
hmmm. If I understand this right, you have a per-fd buffer and this looks wrong. /dev/kmsg is interface of working with the kernel log buffer, why don't we use it?

You need to add a test case which will open /dev/kmsg, write a test message, close /dev/kmsg and then open /dev/kmsg again and check that it will read the test message.

Reply:
Because gvisor doesn't have syslog right now, we (Rahat and I) decided to make this compromise in implementation.

Reply:
https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/kernel/kernel.go;l=1407?q=kernel.syslog&ss=gvisor%2Fgvisor

Pls write a comment which explains in details why we use this buffer and you need to add TODO here.

Reply:
And I am not sure that a per-fd buffer is acceptable even in case of this compromise. What use-case does it cover? Do we have a real app for which this will be useful?

Reply:
hmm, at the beginning of this task I was informed that we don't have syslog and that I should just implement a buffer to store messages. Rahat mentioned that he wants to keep this change simple so that might be why.

As for use cases, this patch is a direct response to [issue 2290](https://github.com/google/gvisor/issues/2290), which only check if /dev/kmsg is there. 

I do agree that adding TODO here would be much better, so I've created an issue and added TODO comment.

Reply:
can we register /dev/null as /dev/kmsg?;)

Reply:
For LTP, this will work with the same result:
https://github.com/linux-test-project/ltp/blob/e566f9ac7816b38bf16ba3570806427385dbf98b/pan/ltp-pan.c#L1370

Reply:
Yup, that would work for  [issue 2290](https://github.com/google/gvisor/issues/2290). Let's see what @mrahatm has to say. Since he suggested me to implement buffer instead of just register /dev/null, so I think he might have some consideration.

Reply:
There are two proposals here:

1. Unblock LTP by registering a null device for /dev/kmsg.
2. Actually implement /dev/kmsg with some basic capabilities.

This patch started off as #2 and it's reasonable to complete it as #2. Andrei is correct that the buffer needs to be global and that we should probably centralize the buffer in the existing syslog.